Add configuration for vert.x with native transport #6669

pcasaes opened the pull request:
Description in issue #6669

cescoffier commented:
LGTM, but would like the POV from @vietj.

cescoffier commented:
Wondering about native actually. Did you test in native mode?

pcasaes commented:
I did a negative test in native (it worked). I didn't actually try to use the _native transport_ in native though. I don't know if that is actually possible.

pcasaes commented:
To be more specific _native transport_ was enabled in native, but I didn't check to see if it was a NOOP.

gastaldi commented:
Is this setting ignored in Windows 10?

gastaldi commented on extensions/netty/runtime/pom.xml:
```diff
@@ -39,6 +39,17 @@
+            <groupId>io.netty</groupId>
+            <artifactId>netty-transport-native-kqueue</artifactId>
+            <classifier>osx-x86_64</classifier>
+        </dependency>
```
Maybe leave these dependencies to be manually added in the project if the property is set? Not sure it makes sense to include the OSX and Linux dependencies in `quarkus-netty`

pcasaes replied:
It's perfectly safe to include these dependencies and they are ignored in Windows. Both Spring Boot and Wildfly have them baked in (the later in netty-all). We could do as yu suggest but then we would have to add some documentation.

pcasaes commented on extensions/netty/runtime/pom.xml:
```diff
@@ -39,6 +39,19 @@
+            <artifactId>netty-transport-native-kqueue</artifactId>
+            <classifier>osx-x86_64</classifier>
+            <optional>true</optional>
+        </dependency>
```
The native transport dependencies are optional. You'll have to add them in manually to your project but the version will be taken care for you.

gastaldi commented on docs/src/main/asciidoc/vertx.adoc:
```diff
@@ -47,6 +47,44 @@ Otherwise, you can manually add this to the dependencies section of your `pom.xm
+== Native Transport
+
+Vert.x is capable of using https://netty.io/wiki/native-transports.html[Netty's native transports] which offers
+performance improvements on certain platforms. To enable them you must include one or both of the following dependencies:
```
How to know when to include one or both? Aren't these dependencies OS specific?

pcasaes replied:
The referenced documentation https://netty.io/wiki/native-transports.html is where you get that information. But it won't break if you include both of them. Netty is smart enough to fallback to a non native transport. Usually projects include both so they'll be platform agnostic.

pcasaes replied:
To be clear. If you include both netty is smart enough to know which one it should use (or non tall) depending on the platform.

gastaldi replied:
Gotcha. Thanks!

pcasaes replied:
I improved the documentation to make it clearer.

gastaldi commented on extensions/vertx-core/runtime/src/test/java/io/quarkus/vertx/core/runtime/VertxCoreProducerTest.java:
```diff
@@ -61,6 +61,7 @@ public void shouldNotFailWithDefaultConfig() {
         configuration.workerPoolSize = 10;
         configuration.warningExceptionTime = Duration.ofSeconds(1);
         configuration.internalBlockingPoolSize = 5;
+        configuration.preferNativeTransport = false;
```
No need to initialize as the default value is false anyway

pcasaes replied:
You're right.

gastaldi commented on extensions/vertx-core/runtime/src/test/java/io/quarkus/vertx/core/runtime/VertxCoreProducerTest.java:
```diff
@@ -100,6 +100,7 @@ private VertxConfiguration createDefaultConfiguration() {
         vc.maxWorkerExecuteTime = Optional.of(Duration.ofSeconds(1));
         vc.internalBlockingPoolSize = 20;
         vc.useAsyncDNS = false;
+        vc.preferNativeTransport = false;
```
No need to initialize here too, I suppose? :)

pcasaes replied:
This method creates a default configuration and sets all values. I figured best to keep it consistent.

vietj reviewed (commented):
A few comments / questions :

- would it be good to log if using native actually worked using `vertx.isNativeTransportEnabled()` ?
- would be interesting to allow binding of the HTTP server on a domain socket path via configuration ?
- it would be interesting to document that Vert.x Web Client can then access domain socket bound servers or give a way in Quarkus to benefit from this

pcasaes commented:
> would it be good to log if using native actually worked using vertx.isNativeTransportEnabled() ?

It certainly would!

> would be interesting to allow binding of the HTTP server on a domain socket path via configuration ?
it would be interesting to document that Vert.x Web Client can then access domain socket bound servers or give a way in Quarkus to benefit from this

It would be, but maybe on a separate ticket/pr?

pcasaes commented:
> would be interesting to allow binding of the HTTP server on a domain socket path via configuration ?

Are you thinking either/or or both domain sockets and tcp?

pcasaes commented:
@vietj I added in the ability to configure a domain socket with documentation on how and why to do this.

pcasaes commented on extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/VertxHttpRecorder.java:
```diff
@@ -541,39 +592,85 @@ public void addRoute(RuntimeValue<Router> router, Function<Router, Route> route,
 
-        private final int port;
-        private final int httpsPort;
-        private final String host;
```
This variables are redundant since their values are in `httpOptions` and `httpsOptions`.

stuartwdouglas commented on extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/HttpConfiguration.java:
```diff
@@ -96,6 +96,36 @@
+     * Enable socket reuse port
+     */
+    @ConfigItem(defaultValue = "false", name = "so-reuse-port")
+    public boolean soReusePort;
```
Why have you changed the default to false? I thought the usual default value here would be true?

pcasaes replied:
It's a new configuration. In this case one that enables [SO_REUSEPORT](https://lwn.net/Articles/542629/). At a first glance I figured it safer to leave this new configurations turned off. 

All the new configurations here only take effect if native transport is enabled. I'll add that to the documentation.

stuartwdouglas replied:
Sorry, I should have read it more closely, I was thinking it was SO_REUSEADDR

pcasaes replied:
No worries. I added this to the docs.

vietj reviewed (approved):
This is looking good and helpful to me.

pcasaes commented:
@cescoffier veitj gave his thumbs up. Can we get this merged?

gastaldi commented:
Can you squash your commits into a single one? Thanks!

pcasaes commented:
Sure

pcasaes committed:
Add configuration for vert.x with native transport #6669

Make Netty's native transport dependencies optional
Improve Netty's native transport documentation
Add Vertx initialization log
Add http configuration for domain sockets
Document http configuration for linux/macos socket options

cescoffier commented on extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/HttpConfiguration.java:
```diff
@@ -35,7 +35,7 @@
      */
     @ConfigItem(defaultValue = "0.0.0.0")
-    public String host;
+    public Optional<String> host;
```
Why switching this property as optional? It has a default value, so the optional will never be empty.

pcasaes replied:
So if you only want to use unix domain sockets you can set
`
quarkus.http.host=
`

And it will disable listening on `host:port`. Do you think it would be clearer to use a different property to disable this?

pcasaes replied:
maybe have remove the `Optional` and add this properties with these default?

```
quarkus.http.host.enabled=true
quarkus.http.host.domain-socket.enabled=false
```

pcasaes replied:
`quarkus.http.host.domain-socket` won't have a default value so it must be optional.

So then we'd only need
```
quarkus.http.host.enabled=true
```

But having `domain-socket` and `http` enabled differently is a bit strange.

pcasaes replied:
Another option is to give `domain-sockets` a default value `/var/run/io.quarkus.app.socket`. That would keep it consistent.

pcasaes replied:
@cescoffier  I went ahead and made these changes.

cescoffier commented on extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/HttpConfiguration.java:
```diff
@@ -96,6 +96,36 @@
+     * Enable socket reuse port (linux/macOs native transport only)
+     */
+    @ConfigItem(defaultValue = "false", name = "so-reuse-port")
+    public boolean soReusePort;
```
Any reason to set the name? The name you use are the default ones.

pcasaes replied:
You're right. I'll fix this.

cescoffier commented on extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/VertxHttpRecorder.java:
```diff
@@ -420,9 +447,13 @@ private static HttpServerOptions createSslOptions(HttpConfiguration httpConfigur
         }
         serverOptions.setSsl(true);
-        serverOptions.setHost(httpConfiguration.host);
+        serverOptions.setHost(httpConfiguration.host.get());
```
no need for `get` if you remove the optional.

cescoffier commented on extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/VertxHttpRecorder.java:
```diff
@@ -345,6 +368,10 @@ public void handle(AsyncResult<Void> event) {
      */
     private static HttpServerOptions createSslOptions(HttpConfiguration httpConfiguration, LaunchMode launchMode)
             throws IOException {
+        if (!httpConfiguration.host.isPresent()) {
```
cannot be false, it has a default value.

cescoffier commented on extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/VertxHttpRecorder.java:
```diff
@@ -497,13 +528,33 @@ private static String findKeystoreFileType(Path storePath) {
 
     private static HttpServerOptions createHttpServerOptions(HttpConfiguration httpConfiguration,
             LaunchMode launchMode, String websocketSubProtocols) {
+        if (!httpConfiguration.host.isPresent()) {
```
cannot be `false`. It has a default value.
In addition, the caller need to be aware of the possibility to get `null`.

cescoffier replied:
It has a `null`-check, so it's ok, but it can still never be `false`.

cescoffier commented on extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/VertxHttpRecorder.java:
```diff
@@ -497,13 +528,33 @@ private static String findKeystoreFileType(Path storePath) {
         // TODO other config properties
         HttpServerOptions options = new HttpServerOptions();
-        options.setHost(httpConfiguration.host);
+        options.setHost(httpConfiguration.host.get());
```
no need for `get` if you remove the optional.

pcasaes committed:
Remove already default explicit config names

pcasaes committed:
Add explicit conf to enable/disable host or domain socket listener

gastaldi commented:
@cescoffier can you review the PR again when possible? Thanks

gsmet commented:
@cescoffier should we make progress on that one?

cescoffier commented:
Everything is fine from my point of view.

cescoffier commented:
Everything is fine from my point of view.
